README.md
bSpokeLight -- custom firmware for LED spoke lights
===================================================

This project provides an alternative firmware for LED spoke lights. Currently,
only the model **YQ8003** (128 LEDs) is supported.

What is the YQ8003?
-------------------

[![See here](https://img.youtube.com/vi/oN4Q87aC5l4/0.jpg)](https://www.youtube.com/watch?v=oN4Q87aC5l4)

Why a new firmware?
-------------------

The advantages of this code over the official one is:

 * Adjustment for the magnet position, so that the displayed images are not
   rotated.
 * Adjustment for the width of the hub. With wide hubs, the original software
   produces a flickering effect with two copies of the picture overlaying each
   other.
 * More convenient loading of pictures. The official software works only on
   Windows, and is very crappy to use. Our software works on all operating
   systems, and is simple to use: Just give it a list of images and display
   durations. It even understands animated gifs.

Some features that are missing compared to the original version (but could be
added):

 * An animationt that looks nice when the weel is not turning, for while the
   bike is stationary.
 * Automatic shut-off.

How do I use it?
----------------

Once you installed it (see below), you run

    $ bSpokeLight --output my-firmware.bin image1.png 10 image2.png 10

to create the firmware `my-firmware.bin` which will display `image1.png` for 10
seconds and `image2.png` for 20 seconds. You can (currently) specify up to 8
images.

The images should be quadratic (otherwise they will be deformed), and ideally
use only the eight colors black, white, red, green, blue, yellow, magenta,
cyan.

To load the firmware onto your wheel, use any STC ISP flash tool, such as
[stcgal](https://github.com/grigorig/stcgal):

    ../path/to/stcgal.py my-firmware.bin

Calibrating your wheel
----------------------

Look at the wheel from the left side. Where is the magnet? Picture the hour arm
of a clock pointing that direction. Which hour is that (e.g. 10)? Pass this
number as the **rotation**.

Now rotate the wheel so that the LED strip is horizontal, and the end with the
white sensor is on the left. How far is the strip above the center of the hub?
Measure this number in “LEDs”, i.e. using another strip. Pass this number as
the **offset**. If the LED strip is actually below the hub, then the number is negative.

You can now pass these parameters when creating your firmware, for example
(with a very wide hub):

    $ bSpokeLight --offset 7 --rotation 10.5 --output my-firmware.bin …


Installation
------------

Currently, the installation is relatively tedious, but until I know of
significant interest, I am avoiding the effort of polishing this.

 * Install `ghc`, `cabal-install`, `binutils` and `sdcc`.
 * Run `make -C firmware` to build the firmware image.
 * Run `cabal install --bindir=.` to build the `bSpokeLight` tool.

You should now find a `bSpokeLight` binary. The firmware is embedded in it, so
you can use it independently from other files.


Hacking on bSpokeLight
----------------------

This is what I learned about the YQ8003 hardware:

 * The microcontroller is a STC12LE5A60S2 with 22MHz and 60KB of RAM. [Data
   sheet](www.stcmcu.com/datasheet/stc/stc-ad-pdf/stc12c5a60s2-english.pdf)

   It is 8051 compatible, so a lot of generic information on how to program
   this microcontroller is online.
 * The magnet triggers external interrupt 0.
 * The LEDs are controlled as follows:

   - `P3_4 = 0` enables the lights.
   - The 8 bits of `P1` indicate which groups of LEDs are adressed. Bits 0,1,2,3
     address the groups on the arm with the buttons, counted from the middle,
     while bits 7,6,5,4 address the groupos on the arm with the sensor, again
     from the middle.
   - `P2`, _negated_, actually sets the LEDs of all addressed groups, with bit
     0 addressing the leds further out and bit 0 the one closest to the middle.
   - All LEDs always share the same color. `P3_5 = 0` is green, `P3_6 = 0` is
     red and `P3_7 = 0` is blue. At most one of these pins should be set to
     zero at a time.

Of course the actual firmware code expects the images in a particular format,
and the `bSpokeLight` program injects the images into the binary; see the code
for my choices there.

Can it support other spoke lights?
----------------------------------

Maybe, if the they are similar enough. Talk to me!

Contact
-------

Please reports bugs and missing features at the [GitHub bugtracker]. This is
also where you can find the [source code].

`bSpokeLight` was written by [Joachim Breitner] and is licensed under a
permissive MIT [license].

[GitHub bugtracker]: https://github.com/nomeata/bSpokeLight/issues
[source code]: https://github.com/nomeata/bSpokeLight
[Joachim Breitner]: http://www.joachim-breitner.de/
[license]: https://github.com/nomeata/bSpokeLight/blob/LICENSE




ChangeLog.md
# Revision history for bSpokeLight

## 0.1 -- unreleased

* First version. Released on an unsuspecting world.


firmware/firmware.c
#include <stc12.h>
#include <stdint.h>

//__bit on = 1;

#define FRAMES 256
#define IMAGES 8
// 256 reloads per second
#define RELOAD_T1  58374
#define RELOAD_T1_HI 0xE4
#define RELOAD_T1_LO 0x06

__code const uint16_t timing[IMAGES] =
	{ 23*256+42,23*256+42,23*256+42,23*256+42,23*256+42,23*256+42,23*256+42,23*256+42 };

__code const uint8_t images[IMAGES][FRAMES * 3 * 8] =
	{"THIS IS WHERE THE IMAGE STARTS"};

uint16_t counter = 0;
uint16_t time10ms = timing[0];
uint8_t image_num = 0;
uint16_t step = 1000000/FRAMES;
uint16_t counter2 = 100000/FRAMES;
uint8_t frame = 0;


void Ext1_Isr() __interrupt(0) __using(2)
{
  //on = 1;
  step = counter/FRAMES;
  counter2 = step;
  counter = 0;
  frame = 0;
}

void ISR_Timer0() __interrupt(1) __using(2)
{
  counter++;
  counter2--;
  if (!counter2) {
	counter2 = step;
	frame++;
  }
  TH0=0xff; // increase timer resolution
  TL0=0x80;
}

void Ext2_Isr() __interrupt(2) __using(2)
{
   IAP_CONTR = 0x60;
}

void ISR_Timer1() __interrupt(3) __using(2)
{
  time10ms--;
  while (!time10ms) {
	image_num++;
	image_num %= 8;
	time10ms = timing[image_num];
  }
  TH1=RELOAD_T1_HI;
  TL1=RELOAD_T1_LO;
}



/*
P1: 8 bits for each group, positive
P2: 8 bits for each LED, negative 

P3_5 = 0: green
P3_6 = 0: red
P3_7 = 0: blue

 */

inline void clear () {
  P2 = 0xff;
  P1 = 0xff;
  P1 = 0x00;
}

// blocks counted from the middle
// leds from the outside, already negated
inline void setlight(const uint8_t *led_ptr) {
	P2 = led_ptr[0]; P1_0 = 1; P1_0 = 0;
	P2 = led_ptr[1]; P1_1 = 1; P1_1 = 0;
	P2 = led_ptr[2]; P1_2 = 1; P1_2 = 0;
	P2 = led_ptr[3]; P1_3 = 1; P1_3 = 0;

	P2 = led_ptr[4]; P1_7 = 1; P1_7 = 0;
	P2 = led_ptr[5]; P1_6 = 1; P1_6 = 0;
	P2 = led_ptr[6]; P1_5 = 1; P1_5 = 0;
	P2 = led_ptr[7]; P1_4 = 1; P1_4 = 0;
}
#define SETLIGHT(image,p) \
	P2 = image[i++]; P1_0 = 1; P1_0 = 0; \
	P2 = image[i++]; P1_1 = 1; P1_1 = 0; \
	P2 = image[i++]; P1_2 = 1; P1_2 = 0; \
	P2 = image[i++]; P1_3 = 1; P1_3 = 0; \
                                             \
	P2 = image[i++]; P1_7 = 1; P1_7 = 0; \
	P2 = image[i++]; P1_6 = 1; P1_6 = 0; \
	P2 = image[i++]; P1_5 = 1; P1_5 = 0; \
	P2 = image[i++]; P1_4 = 1; P1_4 = 0; \

void draw_image() {
	while (1) {
		unsigned char g;
		unsigned int i;
		__code const uint8_t *current_image;

		current_image = images[image_num];
		g = frame;
		// if (g >= FRAMES) { g = FRAMES-1; }
		clear();
		P3_7 = 1; P3_6 = 0; P3_5 = 1;
		i = g*8*3;
		SETLIGHT(current_image,p)

		current_image = images[image_num];
		g = frame;
		// if (g >= FRAMES) { g = FRAMES-1; }
		clear();
		P3_7 = 1; P3_6 = 1; P3_5 = 0;
		i = g*8*3 + 8;
		SETLIGHT(current_image,p)

		current_image = images[image_num];
		g = frame;
		// if (g >= FRAMES) { g = FRAMES-1; }
		clear();
		P3_7 = 0; P3_6 = 1; P3_5 = 1;
		i = g*8*3 + 2*8;
		SETLIGHT(current_image,p)
	}

}

void main () {
	unsigned long int x;
	uint8_t i = 0;

	// activate magnet?
	IT0 = 1;
	EX0 = 1;

	// reset via USB
	IT1 = 1;
	EX1 = 1;

	// Timer
	TMOD = 0x11;
	TR0 = 1;
	ET0 = 1; // enable timer interrupt
	TR1 = 1;
	ET1 = 1;

	TH1=RELOAD_T1_HI;
	TL1=RELOAD_T1_LO;

	// Unclear purpose, later the code reads from P4_2.
	//P4_0 = 1;

	EA = 1; // Enable interrupts

	// Wait a while (to flash it)
	x = 0x20000; while (x--) {};

	P3_4 = 0;
	P2 = 0xff;
	P1 = 0x01;
	draw_image();
}


bSpokeLight.hs
{-# LANGUAGE TupleSections #-}
{-# LANGUAGE TemplateHaskell #-}
{-

Space in the middle: 9
Off-center: 7

-}

import Codec.Picture
import Codec.Picture.Gif
import Data.List.Split
import Data.Complex
import System.Environment
import System.FilePath
import qualified Data.ByteString as BS
import qualified Data.Bits.Bitwise as BA
import Options.Applicative
import Control.Monad
import Data.Monoid
import System.Exit
import Data.Char
import Data.FileEmbed

cFRAMES = 256
cCENTER = 9
cLEN=33+cCENTER/2

data Color = R|G|B deriving (Show, Eq)


type Offset = Double
type Rotation = Double

type BitMaker = (Color -> Bool -> Complex Double -> Bool) -> [Bool]


bitBuilder :: Offset -> Rotation -> BitMaker
bitBuilder off rot f =
    [ f c a (((pos :+ off) * cis rho * cis rot_radians) / (radius :+ 0))
    | frame <- [0,1..cFRAMES-1]
    , let rho = fromIntegral frame*2*pi/ fromIntegral cFRAMES
    , c <- [R,G,B]
    , (a, pos) <- ((False,) <$> arm) ++ (((True,) . negate) <$> arm)
    ]
  where
    rot_radians = (-rot-3)/12*2*pi
    arm = concatMap reverse $ chunksOf 8 $ map (cCENTER/2 +) [0..31]
    radius = sqrt (cLEN^2 + off^2)

getColor :: Color -> PixelRGB8 -> Bool
getColor R (PixelRGB8 r _ _ ) = r > 130
getColor G (PixelRGB8 _ g _ ) = g > 130
getColor B (PixelRGB8 _ _ b ) = b > 130

calibrationBits :: [Bool]
calibrationBits = map not $ foldr1 (zipWith (||)) $
    [ bitBuilder off 0 $ \c a z ->
        abs (phase z - off*pi/10) < pi/40 &&
       (c == oc) &&
       (if a then magnitude z > 0.75
             else (magnitude z < 0.75 && (magnitude z > 0.5 || off == 0)))
    | (oc, off) <- zip (cycle [R,G,B]) [-8..8] ]

dynImageToPackagedData builder dynImage =
    bitsToData $ 
    builder $ \col _ (x :+ y) ->
           let pix = pixelAt pixels
                  (round ((( x+1)/2) * w))
                  (round (((-y+1)/2) * h))
           in not $ getColor col pix
  where
    pixels = convertRGB8 dynImage

    w = fromIntegral (imageWidth pixels)
    h = fromIntegral (imageHeight pixels)

bitsToData = BS.pack . map BA.fromListLE . chunksOf 8


getImage :: BitMaker -> (FilePath, Double) -> IO [(BS.ByteString, Double)]
getImage builder source = case source of
        ("CALIBRATION",n) -> do
            putStrLn $ "Adding offset calibration image"
            pure [(bitsToData calibrationBits, n)]
        (filename, _) | takeExtension filename == ".gif" -> do
            gifData <- BS.readFile filename
            let gifImages = either error id $ decodeGifImages gifData
            let gifDelays = either error id $ getDelaysGifImages gifData
            putStrLn $ "Adding " ++ filename ++ " (" ++ show (length gifImages) ++ " frames)"

            pure $ zip (map (dynImageToPackagedData builder) gifImages)
                       (map ((/100) . fromIntegral) gifDelays)
        (filename, n) -> do
            putStrLn $ "Adding " ++ filename
            dynImage <- either error id <$> readImage filename
            let bits = dynImageToPackagedData builder dynImage
            pure [(bits, n)]

replace :: BS.ByteString -> BS.ByteString -> BS.ByteString -> BS.ByteString
replace pat replacement source
    | BS.null after
    = error $ "replaceIn: Pattern not found: " ++ show pat
    | BS.length replacement > BS.length after
    = error "replaceIn: Replacement larger than remaining string"
    | otherwise = before <> replacement <> BS.drop (BS.length replacement) after
    where
        (before, after) = BS.breakSubstring pat source

template :: BS.ByteString
template = $(embedFile "firmware/firmware.bin")

work :: FilePath -> Offset -> Rotation -> [(FilePath, Double)] -> IO ()
work output offset rotation timed_sources = do
    let builder = bitBuilder offset rotation
    timed_data <- concat <$> mapM (getImage builder) timed_sources
    let (imagesData, timings) = unzip timed_data
    let imageData = BS.concat imagesData

    when (length imagesData > 8) $ do
        putStrLn $ "Too many images"
        exitFailure


    let imagePattern = BS.pack (map (fromIntegral.ord) "THIS IS WHERE THE IMAGE STARTS")
    let timingData    = BS.pack $ map fromIntegral $
            concatMap (\x -> [x`mod`256, x`div`256]) $
            map round $
            map (*256) $
            take 8 $ timings ++ repeat 0
    let timingPattern = BS.pack (take 16 (cycle [42,23]))

    putStrLn $ "Writing " ++ output
    BS.writeFile output $
        replace imagePattern imageData $
        replace timingPattern timingData $
        template

-- Argument handling

main :: IO ()
main = join . customExecParser (prefs showHelpOnError) $
  info (helper <*> parser)
  (  fullDesc
  <> header "YQ3008 image generator"
  -- <> progDesc "Fills a file with as much zeroes as possible"
  )
  where
    parser :: Parser (IO ())
    parser = work
        <$> strOption
            (  long "output"
            <> short 'o'
            <> metavar "FILE"
            <> help "output file"
            )
        <*> option auto
            (  long "offset"
            <> metavar "OFFSET"
            <> help "offset of the bar from the hub (in leds)"
            )
        <*> option auto
            (  long "rotation"
            <> metavar "ROTATION"
            <> help "position of the magnet [0..12]"
            )
        <*> some (
            (,) <$> strArgument
                    (  metavar "IMAGE"
                    <> help "input file (can be CALIBRATION)"
                    )
                <*> argument auto
                    (  metavar "DURATION"
                    <> help "duration (in seconds)"
                    )
            )

